### ISSUE TITLE
Properly deal with non-existent recurrences

### ISSUE THREAD

**Issue by minichma** [rianjs/ical.net#681]
RFC 5545 section 3.3.10 seems to specify conflicting rules regarding how to deal with non-existent recurrences.

1)
    Recurrence rules may generate recurrence instances with an invalid
    date(e.g., February 30) or nonexistent local time(e.g., 1:30 AM
    on a day where the local time is moved forward by an hour at 1:00
    AM).  Such recurrence instances MUST be ignored and MUST NOT be
    counted as part of the recurrence set.

2)
    If the computed local start time of a recurrence instance does not
    exist, or occurs more than once, for the specified time zone, the
    time of the recurrence instance is interpreted in the same manner
    as an explicit DATE-TIME value describing that date and time, as
    specified in Section 3.3.5.

My interpretation would be that 1) applies to 'BY' rules, a rule like

```ics
DTSTART;TZID=Europe/Berlin:20250330T003000
RRULE:FREQ=HOURLY;BYHOUR=0,2,4;COUNT=2
```

would generate `20250330T003000,20250330T043000`, because `20250330T023000` doesn't exist and MUST be ignored.

On the other hand something like 

```ics
DTSTART;TZID=Europe/Berlin:20250329T023000
RRULE:FREQ=DAILY;COUNT=2
```

would generate `20250329T023000,20250330T033000`, because `20250330T023000` must be treated according to section 3.3.5, which requires the previous tz offset to be applied.

According to my interpretation, individual nonexistent `BY` rules would be ignored but whole intervals wouldn't. But that's just a guess at this time.

Consider the following test case. Which values would be expected if DTSTART or DTEND are nonexistent?

```cs
    [TestCase("DTSTART;TZID=Europe/Vienna:20250316T023000", "DTEND;TZID=Europe/Vienna:20250323T023000", "20250316T023000/PT168H", "20250323T023000/PT168H", "20250330T033000/PT168H")]

    // TBD: Which are the right durations in cases where DTSTART or DTEND is non-existing?
    [TestCase("DTSTART;TZID=Europe/Vienna:20250316T023000", "DURATION:P1W",                             "20250316T023000/PT168H", "20250323T023000/PT???", "20250330T033000/PT???")]

    public void DurationOfRecurrencesOverDst(string dtStart, string dtEnd, string d1, string d2, string d3)
    {
        var iCal = Calendar.Load($"""
            BEGIN:VCALENDAR
            BEGIN:VEVENT
            {dtStart}
            {dtEnd}
            RRULE:FREQ=WEEKLY;COUNT=3
            END:VEVENT
            END:VCALENDAR
            """);

        var start = iCal.Events.First().Start;

        var periodSerializer = new PeriodSerializer();
        var expectedPeriods =
            new[] { d1, d2, d3 }
            .Where(x => x != null)
            .Select(x => (Period)periodSerializer.Deserialize(new StringReader(x)))
            .ToArray();

        foreach (var p in expectedPeriods)
        {
            p.StartTime = p.StartTime.ToTimeZone(start.TzId);
            p.EndTime = p.StartTime.Add(p.Duration.Value);
        }

        // date only cannot have a time zone
        EventOccurrenceTest(
            cal: iCal,
            fromDate: null,
            toDate: null,
            expectedPeriods: expectedPeriods,
            timeZones: null,
            eventIndex: 0
        );
    }
```

**Comment by axunonb** [comment]
I share your interpretation, and *Python* with the `datetime` and `pytz` libraries create the same results

**Comment by minichma** [comment]
> I share your interpretation, and _Python_ with the `datetime` and `pytz` libraries create the same results

Oh, good to hear. You mean regarding omitting vs ignoring recurrences, right? I removed the test case from the description though, because I think my interpretation regarding how to adjust nonexistent times could be wrong.

**Comment by axunonb** [comment]
> regarding omitting vs ignoring recurrences

Yes

Although you remove the test case from the description, it would be good to include it in the repo together with the reasoning for this implementation. Very nice finding BTW

**Comment by axunonb** [comment]
Gave your examples another try with *Python* using `icalendar` [library](https://github.com/collective/icalendar) to double-check results. Here we are:
```python
import icalendar
from datetime import datetime
from dateutil.rrule import rrulestr
import pytz

# Create a calendar and event
cal = icalendar.Calendar()
event = icalendar.Event()

# Define the event properties
berlin_tz = pytz.timezone('Europe/Berlin')
dtstart = berlin_tz.localize(datetime(2025, 3, 29, 2, 30))
event.add('dtstart', dtstart)
event.add('rrule', {'FREQ': 'DAILY', 'COUNT': 2})

# Add the event to the calendar
cal.add_component(event)

# Extract the RRULE and DTSTART
rrule = event.get('rrule').to_ical().decode()
dtstart = event.get('dtstart').dt

# Calculate occurrences
rule = rrulestr(rrule, dtstart=dtstart)
occurrences = list(rule)

# Print occurrences
for occ in occurrences:
    print(occ)
```
Output (in both cases):
```
2025-03-29 02:30:00+01:00
2025-03-30 02:30:00+01:00
```
[Edit]
I used some other test cases from us which return obviously wrong occurrences. We should neglect their results.
Still it would be good to have another library for double-checking.

**Comment by minichma** [comment]
Very interesting, thanks for testing! Obviously `2025-03-30 02:30:00+01:00` doesn't exist, so how to resolve this?

* Ignore (aka discard) the entry as required in 1)? Not sure, but rather not.
* Move to `2025-03-30 03:30:00+02:00` as required by 2)? Probably.
* Anything else, e.g. `2025-03-30 03:00:00+02:00`? Probably not.

**Comment by axunonb** [comment]
Yes, let's stick to "our way". 
I also had a brief look at other Python libraries, and I think ical.net is meanwhile doing a pretty good job on recurrence and covering RFC 5545 in general.

**Comment by axunonb** [comment]
@minichma This was resolved with #680?

**Comment by minichma** [comment]
@axunonb I don't think so. #680 improved the situation but I don't think the problem has been fully addressed. The cases with `FREQ` leading to nonexistent instants is covered by tests `DurationOfRecurrencesOverDst` and works as described above, but the case with `BY` rules leading to nonexistent isn't covered and doesn't work, at least not as discussed above.

**Comment by minichma** [comment]
Reconsidering the case again, I think this is a mistake in RFC 5545. I believe that 1) doesn't apply to DST changes, only to non-existent dates

> Recurrence rules may generate recurrence instances with an invalid
date(e.g., February 30) or nonexistent local time(e.g., 1:30 AM
on a day where the local time is moved forward by an hour at 1:00
AM). Such recurrence instances MUST be ignored and MUST NOT be
counted as part of the recurrence set.

That's also how Google Calendar deals with the case. The following events would both produce the same recurrences:

* 20250329T023000
* 20250330T033000
* 20250331T033000

```ics
BEGIN:VEVENT
DTSTART;TZID=Europe/Rome:20250329T023000
DTEND;TZID=Europe/Rome:20250329T033000
RRULE:FREQ=DAILY;COUNT=3
DTSTAMP:20250213T094748Z
...
END:VEVENT
```

```ics
BEGIN:VEVENT
DTSTART;TZID=Europe/Rome:20250329T023000
DTEND;TZID=Europe/Rome:20250329T033000
RRULE:FREQ=WEEKLY;WKST=MO;COUNT=3;BYDAY=SA,SU,MO
...
END:VEVENT
```

I think that's the right approach and that the specification from the RFC _'or nonexistent local time(e.g., 1:30 AM
on a day where the local time is moved forward by an hour at 1:00
AM)'_ should be ignored in favor of what is written in 2).
